core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,195 @@
+    private volatile boolean closed;
+
+    private StreamMessageInputStream(StreamMessage<T> source,
+                                     Function<? super T, ? extends HttpData> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        this.source = source;
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    private int read(Function<InputStream, Integer> function) throws IOException {
+        ensureOpen();
+        ensureSubscribed();
+        if (inputStream == null || inputStream.available() == 0) {
+            try {
+                inputStream = subscriber.nextStream();
+            } catch (InterruptedException e) {
+                final InterruptedIOException ioe = new InterruptedIOException();
+                ioe.initCause(e);
+                throw ioe;
+            }
+        }
+        return function.apply(inputStream);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        source.abort();
+        if (inputStream != null) {
+            inputStream.close();
+        }
+    }
+
+    @Override
+    public int available() throws IOException {
+        ensureOpen();
+        if (inputStream == null) {
+            return 0;
+        }
+        return inputStream.available();
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureSubscribed() {
+        if (subscribed.compareAndSet(false, true)) {
+            source.subscribe(subscriber);
+        }
+        subscriber.whenSubscribed.join();
+    }
+
+    private static final class StreamMessageInputStreamSubscriber<T> implements Subscriber<T> {
+
+        private static final InputStream EMPTY_STREAM = new InputStream() {
+            @Override
+            public int read() {
+                return -1;
+            }
+        };
+
+        private final Function<? super T, ? extends HttpData> httpDataConverter;
+        @Nullable
+        private Subscription upstream;
+        private final CompletableFuture<Void> whenSubscribed = new CompletableFuture<>();
+        private final BlockingQueue<InputStream> queue = new LinkedBlockingDeque<>();
```

Comment:
`ByteBuffersBackedInputStream` seems to handle much of the logic for presenting an input stream, would it make sense to reduce the logic here by moving that class to core and using it directly here?

https://github.com/line/armeria/blob/master/resteasy%2Fsrc%2Fmain%2Fjava%2Fcom%2Flinecorp%2Farmeria%2Finternal%2Fcommon%2Fresteasy%2FByteBuffersBackedInputStream.java

Instead of `queue.add`, just `byteBuffersStream.add` and instead of EMPTY_STREAM, `setEos`

Reply:
> moving `ByteBuffersBackedInputStream` class to `core` and using it directly here?

Hmm.. I think it make sense! 
But I'm not sure about moving `ByteBuffersBackedInputStream` to `core` module is right way.

WDYT @ikhoon nim? I'll respect maintainer's opinion 😄

Reply:
That is a good idea.
Let's rename `ByteBuffersBackedInputStream` to `ByteBufsInputStream` and place in `internal.common` of `core`.

Reply:
> Let's rename ByteBuffersBackedInputStream to ByteBufsInputStream and place in internal.common of core.

I'll update this PR with this way!

Reply:
> 1. Rename ByteBuffersBackedInputStream to ByteBufsInputStream
> 2. Place in internal.common of core.

✅ updated.